### ISSUE TITLE
Does it converge?

### ISSUE THREAD

**Issue by Bigpig4396** [ajlangley/cpo-pytorch#6]
I transplanted this code to my environment but it did not converge.

**Comment by ajlangley** [comment]
Did you close this issue because it converged after all? Just curious!

**Comment by Bigpig4396** [comment]
It is hard to say, but it shows to converge but corrupted halfway.

**Comment by Bigpig4396** [comment]
several questions about the code

1.
reward_advs -= reward_advs.mean()
reward_advs /= reward_advs.std()
cost_advs -= reward_advs.mean()
cost_advs /= cost_advs.std()

I don't think it should be done. It breaks the true advantage.

2.
reward_loss = -torch.mean(imp_sampling * reward_advs)
constraint_loss = torch.sum(imp_sampling * constraint_advs) / self.simulator.n_trajectories

you see here the reward loss and constraint_loss have different sign + and -. According to the paper, there should be the same. (unless you have negative cost?). so what should be the sign of cost signal from environment? (I suppose it should be positive).

3.
if is_feasible:
      lam, nu = self.calc_dual_vars(q, r, s, c)
      search_dir = -lam ** -1 * (F_inv_g + nu * F_inv_b)

according to the paper, this should be +lam not -lam right?

**Comment by Bigpig4396** [comment]
also i see all the places that self.max_kl should be,  you used 2 * self.max_kl, I don't know why

**Comment by ajlangley** [comment]
I haven’t really maintained this code base in a long time and can’t really answer most of your questions. But I think your questions about the signs are because I interpret cost as a quantity we want to minimize.

**Comment by Bigpig4396** [comment]
So the cost should be a positive scalar correct?

**Comment by ajlangley** [comment]
Correct

**Comment by ajlangley** [comment]
Well, not necessarily positive. To be precise, all that matters is that higher cost is worse.

**Comment by Bigpig4396** [comment]
ok, I see
